sonarqube/resource_sonarqube_qualitygate_usergroup_association.go

```diff
@@ -0,0 +1,176 @@
+package sonarqube
+
+import (
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"net/url"
+	"strings"
+
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+	"github.com/satori/uuid"
+)
+
+// GetQualityGateUsergroupAssociation for unmarshalling response body from getting quality gate association
+type GetQualityGateUsergroupAssociation struct {
+	Paging Paging                                       `json:"paging"`
+	Groups []GetQualityGateUsergroupAssociationProjects `json:"groups,omitempty"`
+	Users  []GetQualityGateUsergroupAssociationProjects `json:"users,omitempty"`
+}
+
+// GetQualityGateUsergroupAssociationProjects used in GetQualityGateUsergroupAssociation
+type GetQualityGateUsergroupAssociationProjects struct {
+	Login       string `json:"login,omitempty"`
+	Name        string `json:"name"`
+	Description string `json:"description,omitempty"`
+	Selected    bool   `json:"selected"`
+}
+
+// Returns the resource represented by this file.
+func resourceSonarqubeQualityGateUsergroupAssociation() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceSonarqubeQualityGateUsergroupAssociationCreate,
+		Read:   resourceSonarqubeQualityGateUsergroupAssociationRead,
+		Delete: resourceSonarqubeQualityGateUsergroupAssociationDelete,
+
+		// Define the fields of this schema.
+		Schema: map[string]*schema.Schema{
+			"login_name": {
+				Type:         schema.TypeString,
+				Optional:     true,
+				ForceNew:     true,
+				ExactlyOneOf: []string{"login_name", "group_name"},
+			},
+			"group_name": {
+				Type:         schema.TypeString,
+				Optional:     true,
+				ForceNew:     true,
+				ExactlyOneOf: []string{"login_name", "group_name"},
+			},
+			"gatename": {
+				Type:     schema.TypeString,
+				Optional: true,
+				ForceNew: true,
+			},
+		},
+	}
+}
+
+func resourceSonarqubeQualityGateUsergroupAssociationCreate(d *schema.ResourceData, m interface{}) error {
+	sonarQubeURL := m.(*ProviderConfiguration).sonarQubeURL
+	rawQuery := url.Values{
+		"gateName": []string{d.Get("gatename").(string)},
+	}
+	target, ok := d.GetOk("login_name")
+
+	if ok {
+		sonarQubeURL.Path = strings.TrimSuffix(sonarQubeURL.Path, "/") + "/api/qualitygates/add_user"
+		rawQuery.Add("login", target.(string))
+	} else {
+		sonarQubeURL.Path = strings.TrimSuffix(sonarQubeURL.Path, "/") + "/api/qualitygates/add_group"
+		rawQuery.Add("groupName", target.(string))
```

Comment:
tests are failing because target is empty here for the case of groupName. Partially my fault from my previous comments. However you want to fix this is good with me. Below is a example. tests should be running on commits so you can check that your change is passing acceptance tests after pushing the fix

```go
	if _, ok := d.GetOk("login_name") {
		sonarQubeURL.Path = strings.TrimSuffix(sonarQubeURL.Path, "/") + "/api/qualitygates/add_user"
		rawQuery.Add("login", d.Get("login_name").(string))
	} else {
		sonarQubeURL.Path = strings.TrimSuffix(sonarQubeURL.Path, "/") + "/api/qualitygates/add_group"
		rawQuery.Add("groupName", d.Get("group_name").(string))
	}
```

Reply:
Once the builds are passing, can you @ me so i remember to merge this. Thanks!

Reply:
i'm working from 3G mobile, so cant test it locally, thanks will use the ci